contracts/implementation/InvariantCheck.sol

```diff
@@ -0,0 +1,58 @@
+//SPDX-License-Identifier: CC-BY-NC-ND-4.0
+pragma solidity 0.8.7;
+
+import "../interfaces/IInvariantCheck.sol";
+import "../interfaces/IPoolCommitter.sol";
+import "../interfaces/IPausable.sol";
+import "../interfaces/ILeveragedPool.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/// @title The contract for checking invariants and pausing if necessary
+/// @notice Every time certain functions are called, known invariants are checked and if any do not hold, contracts are paused.
+contract InvariantCheck is IInvariantCheck {
+    /**
+     * @notice Checks all invariants, and pauses all contracts if
+     *         any invariant does not hold.
+     * @dev This should be called before onlyUnpaused, in case contracts are paused then pause check must happen after.
+     */
+    function checkInvariants(address poolToCheck) external override {
+        ILeveragedPool pool = ILeveragedPool(poolToCheck);
+        IPoolCommitter poolCommitter = IPoolCommitter(pool.poolCommitter());
+        uint256 poolBalance = IERC20(pool.quoteToken()).balanceOf(poolToCheck);
+        (
+            IPoolCommitter.Commitment memory totalMostRecentCommits,
+            IPoolCommitter.Commitment memory totalNextIntervalCommit
+        ) = poolCommitter.getPendingCommits();
+        uint256 pendingMints;
+        unchecked {
+            pendingMints =
+                totalMostRecentCommits.longMintAmount +
+                totalMostRecentCommits.shortMintAmount +
+                totalNextIntervalCommit.longMintAmount +
+                totalNextIntervalCommit.shortMintAmount;
+        }
+        uint256 longBalance = pool.longBalance();
+        uint256 shortBalance = pool.shortBalance();
+        if (!balanceInvariant(poolBalance, pendingMints, longBalance, shortBalance)) {
+            pauseAll(IPausable(poolToCheck), IPausable(address(poolCommitter)));
+        }
+    }
```

Comment:
Fairly sure this isn't safe.

Take the case where a user passes in a `poolToCheck` that is some random contract that they deployed, but it has a real `poolCommitter` contract returned by calling the `pool.poolCommitter()` function.

In this case, a user could craft their contract to ensure the invariant does not hold, and the real `poolCommitter` gets paused, potentially stopping another real pool from working.

The fix would be to validate that the pool passed in is a real pool address as deployed by the factory.

Reply:
Legendary catch